features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/resources/ResourcesDataStoreManager.kt

```diff
@@ -0,0 +1,120 @@
+
+internal class ResourcesDataStoreManager(
+    private val featureSdkCore: FeatureSdkCore,
+    private val resourceHashesSerializer: Serializer<ResourceHashesEntry>,
+    private val resourcesHashesDeserializer: Deserializer<String, ResourceHashesEntry>,
+    private val featureScope: FeatureScope? = featureSdkCore.getFeature(
+        SESSION_REPLAY_RESOURCES_FEATURE_NAME
+    )
+) {
+    private val knownResources = Collections.synchronizedSet(mutableSetOf<String>())
+    private var storedLastUpdateDateNs = System.nanoTime()
+
+    init {
+        initializeManager()
+    }
+
+    internal fun wasResourcePreviouslySent(resourceHash: String): Boolean =
+        knownResources.contains(resourceHash)
+
+    internal fun store(resourceHash: String) {
+        knownResources.add(resourceHash)
+        writeResourcesToStore()
+    }
+
+    // region internal
+
+    private fun initializeManager() {
+        fetchStoredResourceHashes(
+            onFetchSuccessful = { storedEntry ->
+                val lastUpdateDateNs = storedEntry?.data?.lastUpdateDateNs
+                val storedHashes = storedEntry?.data?.resourceHashes
+
+                lastUpdateDateNs?.let {
+                    storedLastUpdateDateNs = lastUpdateDateNs.toLong()
+                }
+
+                if (didDataStoreExpire()) {
+                    handleDataStoreExpired(storedHashes?.toSet())
+                } else {
+                    storedHashes?.let { hashes ->
+                        knownResources.addAll(hashes)
+                    }
+                }
+            },
+            onFetchFailure = {
+                deleteStoredHashesEntry()
+            }
+        )
+    }
+
+    private fun writeResourcesToStore() {
+        val data = ResourceHashesEntry(
+            lastUpdateDateNs = storedLastUpdateDateNs,
+            resourceHashes = knownResources.toList()
+        )
+
+        featureScope?.dataStore?.setValue(
+            data = data,
+            key = DATASTORE_HASHES_ENTRY_NAME,
+            serializer = resourceHashesSerializer
+        )
+    }
+
+    private fun fetchStoredResourceHashes(
+        onFetchSuccessful: (dataStoreContent: DataStoreContent<ResourceHashesEntry>?) -> Unit = {},
+        onFetchFailure: () -> Unit = {}
+    ) {
+        featureScope?.dataStore?.value(
+            key = DATASTORE_HASHES_ENTRY_NAME,
+            deserializer = resourcesHashesDeserializer,
+            callback = object : DataStoreCallback<ResourceHashesEntry> {
+                override fun onSuccess(dataStoreContent: DataStoreContent<ResourceHashesEntry>?) {
+                    onFetchSuccessful(dataStoreContent)
+                }
+
+                override fun onFailure() {
+                    onFetchFailure()
+                }
+            }
+        )
+    }
+
+    private fun handleDataStoreExpired(storedHashes: Set<String>?) {
+        storedHashes?.let { knownResources.removeAll(it) }
+        storedLastUpdateDateNs = System.nanoTime()
```

Comment:
this property can be modified and read from the different threads, so you need to make it `@Volatile` or use atomic reference.

Reply:
I wasn't sure about this because it's only modified in the initialization of the class and otherwise read, and we're using a single instance

Reply:
it still will be accessed probably by the different threads, so if you want to have thread-safe design of this class, it should be compliant